File: app/src/main/java/cm/aptoide/pt/download/FileDownloadManager.java
package cm.aptoide.pt.download;

import cm.aptoide.pt.downloadmanager.Constants;
import cm.aptoide.pt.downloadmanager.FileDownloadCallback;
import cm.aptoide.pt.downloadmanager.FileDownloader;
import com.liulishuo.filedownloader.BaseDownloadTask;
import rx.Completable;
import rx.Observable;

/**
 * Created by filipegoncalves on 7/31/18.
 */

public class FileDownloadManager implements FileDownloader {

  public static final int RETRY_TIMES = 3;
  public static final int PROGRESS_MAX_VALUE = 100;
  private static final int APTOIDE_DOWNLOAD_TASK_TAG_KEY = 888;
  private final String mainDownloadPath;
  private final int fileType;
  private final String packageName;
  private final int versionCode;
  private final String fileName;
  private com.liulishuo.filedownloader.FileDownloader fileDownloader;
  private FileDownloadTask fileDownloadTask;
  private String downloadsPath;
  private int downloadId;

  public FileDownloadManager(com.liulishuo.filedownloader.FileDownloader fileDownloader,
      FileDownloadTask fileDownloadTask, String downloadsPath, String mainDownloadPath,
      int fileType, String packageName, int versionCode, String fileName) {
    this.fileDownloader = fileDownloader;
    this.fileDownloadTask = fileDownloadTask;
    this.downloadsPath = downloadsPath;
    this.mainDownloadPath = mainDownloadPath;
    this.fileType = fileType;
    this.packageName = packageName;
    this.versionCode = versionCode;
    this.fileName = fileName;
  }

  @Override public Completable startFileDownload() {
    return Completable.fromAction(() -> {
      if (mainDownloadPath == null || mainDownloadPath.isEmpty()) {
        throw new IllegalArgumentException("The url for the download can not be empty");
      } else {
        createBaseDownloadTask(mainDownloadPath, versionCode, packageName, fileType, fileName);
        fileDownloader.start(fileDownloadTask, true);
      }
    });
  }

  @Override public Completable pauseDownload() {
    return Completable.fromAction(() -> fileDownloader.pause(fileDownloadTask));
  }

  @Override public Completable removeDownloadFile() {
    return Completable.fromAction(() -> fileDownloader.clear(downloadId, mainDownloadPath));
  }

  @Override public Observable<FileDownloadCallback> observeFileDownloadProgress() {
    return fileDownloadTask.onDownloadStateChanged();
  }

  //the listener is removed to not receive the paused callback
  @Override public void stopFailedDownload() {
    int taskId = fileDownloader.replaceListener(downloadId, null);
    if (taskId != 0) {
      fileDownloader.pause(taskId);
    }
  }

  private void createBaseDownloadTask(String mainDownloadPath, int versionCode, String packageName,
      int fileType, String fileName) {

    BaseDownloadTask baseDownloadTask = fileDownloader.create(mainDownloadPath);
    baseDownloadTask.setAutoRetryTimes(RETRY_TIMES);

    baseDownloadTask.addHeader(Constants.VERSION_CODE, String.valueOf(versionCode));
    baseDownloadTask.addHeader(Constants.PACKAGE, packageName);
    baseDownloadTask.addHeader(Constants.FILE_TYPE, String.valueOf(fileType));
    baseDownloadTask.setTag(APTOIDE_DOWNLOAD_TASK_TAG_KEY, fileDownloadTask);
    baseDownloadTask.setListener(fileDownloadTask);
    baseDownloadTask.setCallbackProgressTimes(PROGRESS_MAX_VALUE);
    baseDownloadTask.setPath(downloadsPath + fileName);
    this.downloadId = baseDownloadTask.asInQueueTask()
        .enqueue();
  }
}

File: aptoide-installer/src/main/java/cm/aptoide/pt/aptoide_installer/data/download/filedownloader/FileDownloadManager.kt
package cm.aptoide.pt.aptoide_installer.data.download.filedownloader

import cm.aptoide.pt.downloadmanager.Constants
import cm.aptoide.pt.downloadmanager.FileDownloadCallback
import cm.aptoide.pt.downloadmanager.FileDownloader
import io.reactivex.Completable
import io.reactivex.Observable

class FileDownloadManager(
  private val fileDownloader: com.liulishuo.filedownloader.FileDownloader,
  private val fileDownloadTask: FileDownloadTask,
  private val downloadsPath: String,
  private val mainDownloadPath: String,
  private val fileType: Int,
  private val packageName: String,
  private val versionCode: Int,
  private val fileName: String
) : FileDownloader {

  companion object {
    const val RETRY_TIMES = 3
    const val PROGRESS_MAX_VALUE = 100
    private const val APTOIDE_DOWNLOAD_TASK_TAG_KEY = 888
  }

  private var downloadId = 0


  override fun startFileDownload(): Completable {
    return Completable.fromAction {
      if (mainDownloadPath.isNullOrEmpty()) {
        throw IllegalArgumentException("The url for the download can not be empty")
      } else {
        setupBaseDownloadTask(mainDownloadPath, versionCode, packageName, fileType, fileName)
        fileDownloader.start(fileDownloadTask, true)
      }
    }
  }

  override fun removeDownloadFile(): Completable {
    return Completable.fromAction {
      fileDownloader.clear(
        downloadId,
        mainDownloadPath
      )
    }
  }

  override fun observeFileDownloadProgress(): Observable<FileDownloadCallback> {
    return fileDownloadTask.onDownloadStateChanged()
  }

  override fun stopFailedDownload() {
    val taskId = fileDownloader.replaceListener(downloadId, null)
    if (taskId != 0) {
      fileDownloader.pause(taskId)
    }
  }

  private fun setupBaseDownloadTask(
    mainDownloadPath: String, versionCode: Int, packageName: String,
    fileType: Int, fileName: String
  ) {
    val baseDownloadTask = fileDownloader.create(mainDownloadPath)
    baseDownloadTask.autoRetryTimes = RETRY_TIMES
    baseDownloadTask.addHeader(Constants.VERSION_CODE, versionCode.toString())
    baseDownloadTask.addHeader(Constants.PACKAGE, packageName)
    baseDownloadTask.addHeader(Constants.FILE_TYPE, fileType.toString())
    baseDownloadTask.setTag(APTOIDE_DOWNLOAD_TASK_TAG_KEY, fileDownloadTask)
    baseDownloadTask.listener = fileDownloadTask
    baseDownloadTask.callbackProgressTimes = PROGRESS_MAX_VALUE
    baseDownloadTask.path = downloadsPath + fileName
    downloadId = baseDownloadTask.asInQueueTask()
      .enqueue()
  }
}